File: Kevin-Muthiani/Arrays_and_Strings/minimum_swaps.py
# https://www.hackerrank.com/challenges/minimum-swaps-2/problem
# You are given an unordered array consisting of consecutive integers 
# [1, 2, 3, ..., n] without any duplicates. You are allowed to swap any 
# two elements. Find the minimum number of swaps required to sort the 
# array in ascending order.


def minimumSwaps(arr):
    swaps = 0
    n = len(arr)
    valueIndexDict = dict(zip(arr, range(1, n+1)))
    
    for i in range(1,n+1):
        if valueIndexDict[i] != i:
            # Swap
            valueIndexDict[arr[i-1]] =  valueIndexDict[i]
            arr[valueIndexDict[i]-1] = arr[i-1]
            swaps += 1    
    
    return swaps


# Test function
if __name__ == "__main__":
    myArray = [4, 3, 1, 2]
    print(minimumSwaps(myArray))

File: arrays-and-strings/MinimumSwaps.js
var a = [7, 1, 3, 2, 4, 5, 6];
var b = [4, 3, 1, 2];
var c = [2, 3, 4, 1, 5];
var swaps = 0;

function minimumSwaps(arr){

    var isSwapped = false;

    for (var i = 0, len = arr.length; i < len; i ++){
        for (var j = 0; j < (len - i - 1); j ++){
            if (arr[j] > arr[j + 1]){
                swap(arr, j, j + 1);
                isSwapped = true;
                swaps++;
            }
        }

        if(!isSwapped){
            break;
        }
    }

    function swap (arr, position, next){
        [arr[position], arr[next]] = [arr[next], arr[position]];
    }

}

console.log(a);

minimumSwaps(a);

console.log(a);

console.log("no of swaps are: " + swaps);